File: Archivos/APE-5/Ape5Ejercicio6.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <sstream>

using namespace std;

// ============================================================================
// FUNCIONES AUXILIARES DE VALIDACIÓN
// ============================================================================

// Valida de forma básica que la fecha tenga el formato numérico DD/MM/AAAA
bool validarFecha(const string& fecha) {
    if (fecha.length() != 10) return false;
    if (fecha[2] != '/' || fecha[5] != '/') return false;
    
    // Verifica que el resto sean números
    for (int i = 0; i < 10; i++) {
        if (i != 2 && i != 5) {
            if (fecha[i] < '0' || fecha[i] > '9') return false;
        }
    }
    return true;
}

// Solicita la fecha y fuerza al ciclo a repetirse hasta que sea válida
string solicitarFecha() {
    string fecha;
    while (true) {
        cout << "  Ingrese fecha (DD/MM/AAAA): ";
        cin >> fecha;
        if (validarFecha(fecha)) {
            return fecha;
        }
        cout << "  [Error] Formato de fecha incorrecto. Use DD/MM/AAAA.\n";
    }
}

// Fuerza al usuario a ingresar únicamente 'A' o 'F'
char solicitarEstado() {
    char estado;
    while (true) {
        cout << "  Ingrese estado (A = Asistio, F = Falto): ";
        cin >> estado;
        estado = toupper(estado); // Lo convierte a mayúscula automáticamente
        if (estado == 'A' || estado == 'F') {
            return estado;
        }
        cout << "  [Error] Estado invalido. Solo se permite A o F.\n";
    }
}

// ============================================================================
// FUNCIONES DE CONTROL DE ASISTENCIA (MANEJO DE ARCHIVOS)
// ============================================================================

// REQUERIMIENTO 1: Registrar nombre, fecha y estado en el archivo
void registrarAsistencia() {
    // ios::app abre el archivo en modo "append" para añadir texto al final sin borrar lo anterior
    ofstream archivoEscritura("asistencia.txt", ios::app);
    
    if (!archivoEscritura.is_open()) {
        cout << "Error: No se pudo abrir el archivo para guardar datos.\n";
        return;
    }

    string nombre, fecha;
    char estado;

    cout << "\n--- NUEVO REGISTRO DE ASISTENCIA ---\n";
    cout << "  Nombre del estudiante: ";
    cin.ignore(); // Limpia el buffer antes de leer una cadena completa con espacios
    getline(cin, nombre);
    
    fecha = solicitarFecha();
    estado = solicitarEstado();

    // Guardamos los datos separados por un caracter delimitador (pipe '|') para facilitar la lectura posterior
    archivoEscritura << nombre << "|" << fecha << "|" << estado << "\n";
    archivoEscritura.close();

    cout << "\n>> Asistencia registrada con exito en el sistema.\n";
}

// REQUERIMIENTO 2 y 3: Mostrar historial completo y contar las faltas totales
void mostrarHistorialYFaltas() {
    ifstream archivoLectura("asistencia.txt");
    
    if (!archivoLectura.is_open()) {
        cout << "\n[Aviso] No hay registros previos de asistencia (El archivo aun no existe).\n";
        return;
    }

    string linea;
    int totalFaltas = 0;
    int totalRegistros = 0;

    cout << "\n=========================================================\n";
    cout << "              HISTORIAL DE ASISTENCIA\n";
    cout << "=========================================================\n";
    cout << "  %-20s %-15s %-10s\n";
    cout << "  -------------------------------------------------------\n";

    // Ciclo para leer el archivo línea por línea hasta el final
    while (getline(archivoLectura, linea)) {
        if (linea.empty()) continue;

        stringstream ss(linea);
        string nombre, fecha, estadoStr;

        // Descomponemos la línea usando el delimitador '|'
        getline(ss, nombre, '|');
        getline(ss, fecha, '|');
        getline(ss, estadoStr, '|');

        char estado = estadoStr[0];

        // Traducimos el caracter plano a un texto legible para el usuario
        string textoEstado = (estado == 'A') ? "Asistio" : "Falto";

        // Si es una falta, incrementamos el contador
        if (estado == 'F') {
            totalFaltas++;
        }

        // Imprimimos la fila formateada de forma ordenada
        cout << "  " << nombre;
        for (size_t i = nombre.length(); i < 22; i++) cout << " ";
        cout << fecha << "         " << textoEstado << "\n";
        
        totalRegistros++;
    }

    archivoLectura.close();

    cout << "=========================================================\n";
    cout << " Total de registros procesados: " << totalRegistros << "\n";
    cout << " Total de FALTAS contabilizadas: " << totalFaltas << "\n";
    cout << "=========================================================\n";
}

// ============================================================================
// PROGRAMA PRINCIPAL
// ============================================================================
int main() {
    int opcion;

    // Ciclo principal del menú de control de asistencia
    do {
        cout << "\n==================================\n";
        cout << "    SISTEMA CONTROL DE ASISTENCIA \n";
        cout << "==================================\n";
        cout << "1. Registrar Asistencia\n";
        cout << "2. Mostrar Historial y Faltas\n";
        cout << "3. Salir del programa\n";
        cout << "----------------------------------\n";
        cout << "Seleccione una opcion (1-3): ";
        
        // Validación para evitar bucles infinitos si ingresan letras en el menú
        if (!(cin >> opcion)) {
            cout << "Opcion no valida. Intente de nuevo.\n";
            cin.clear();
            cin.ignore(10000, '\n');
            continue;
        }

        switch (opcion) {
            case 1:
                registrarAsistencia();
                break;
            case 2:
                mostrarHistorialYFaltas();
                break;
            case 3:
                cout << "\nSaliendo del sistema de asistencia... ¡Hasta luego!\n";
                break;
            default:
                cout << "[Error] La opcion elegida no existe. Intente con 1, 2 o 3.\n";
        }
    } while (opcion != 3); // Rompe el ciclo cuando la opción es 3

    return 0;
}

File: Archivos/APE-5/Ape5Ejercicio6.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.Scanner;

// ============================================================================
// CLASE PRINCIPAL: SistemaAsistencia
// ============================================================================
public class Ape5Ejercicio6 {

    private static final String NOMBRE_ARCHIVO = "asistencia.txt";
    private static final Scanner teclado = new Scanner(System.in);

    // Valida de forma básica que la fecha tenga el formato numérico DD/MM/AAAA
    public static boolean validarFecha(String fecha) {
        if (fecha.length() != 10) return false;
        if (fecha.charAt(2) != '/' || fecha.charAt(5) != '/') return false;
        
        // Verifica que los demás caracteres sean dígitos
        for (int i = 0; i < 10; i++) {
            if (i != 2 && i != 5) {
                if (fecha.charAt(i) < '0' || fecha.charAt(i) > '9') return false;
            }
        }
        return true;
    }

    // Solicita la fecha y fuerza al ciclo a repetirse hasta que sea válida
    public static String solicitarFecha() {
        String fecha;
        while (true) {
            System.out.print("  Ingrese fecha (DD/MM/AAAA): ");
            fecha = teclado.next();
            if (validarFecha(fecha)) {
                return fecha;
            }
            System.out.println("  [Error] Formato de fecha incorrecto. Use DD/MM/AAAA.");
        }
    }

    // Fuerza al usuario a ingresar únicamente 'A' o 'F'
    public static char solicitarEstado() {
        String entrada;
        while (true) {
            System.out.print("  Ingrese estado (A = Asistio, F = Falto): ");
            entrada = teclado.next().toUpperCase(); // Convierte a mayúscula automáticamente
            if (entrada.length() == 1 && (entrada.charAt(0) == 'A' || entrada.charAt(0) == 'F')) {
                return entrada.charAt(0);
            }
            System.out.println("  [Error] Estado invalido. Solo se permite A o F.");
        }
    }

    // REQUERIMIENTO 1: Registrar nombre, fecha y estado en el archivo
    public static void registrarAsistencia() {
        // El parámetro true en FileWriter habilita el modo "append" (añadir al final)
        try (FileWriter fw = new FileWriter(NOMBRE_ARCHIVO, true);
             PrintWriter archivoEscritura = new PrintWriter(fw)) {
            
            System.out.println("\n--- NUEVO REGISTRO DE ASISTENCIA ---");
            System.out.print("  Nombre del estudiante: ");
            teclado.nextLine(); // Limpia el buffer del Scanner
            String nombre = teclado.nextLine();
            
            String fecha = solicitarFecha();
            char estado = solicitarEstado();

            // Guardamos los datos separados por la barra vertical '|'
            archivoEscritura.println(nombre + "|" + fecha + "|" + estado);
            System.out.println("\n>> Asistencia registrada con exito en el archivo.");
            
        } catch (IOException e) {
            System.out.println("Error al intentar escribir en el archivo: " + e.getMessage());
        }
    }

    // REQUERIMIENTO 2 y 3: Mostrar historial completo y contar las faltas totales
    public static void mostrarHistorialYFaltas() {
        File archivo = new File(NOMBRE_ARCHIVO);
        
        // Comprobamos de antemano si el archivo físico existe en el disco
        if (!archivo.exists()) {
            System.out.println("\n[Aviso] No hay registros previos de asistencia (El archivo aun no existe).");
            return;
        }

        int totalFaltas = 0;
        int totalRegistros = 0;

        System.out.println("\n=========================================================");
        System.out.println("              HISTORIAL DE ASISTENCIA");
        System.out.println("=========================================================");
        System.out.printf("  %-22s %-15s %-10s\n", "Estudiante", "Fecha", "Estado");
        System.out.println("  -------------------------------------------------------");

        try (FileReader fr = new FileReader(archivo);
             BufferedReader archivoLectura = new BufferedReader(fr)) {
            
            String linea;
            // Ciclo para leer el archivo línea por línea hasta llegar al final (null)
            while ((linea = archivoLectura.readLine()) != null) {
                if (linea.trim().isEmpty()) continue;

                // Dividimos la cadena usando la barra vertical como separador
                // Usamos \\| porque en Java split requiere una expresión regular y el | es un caracter reservado
                String[] partes = linea.split("\\|");
                
                if (partes.length == 3) {
                    String nombre = partes[0];
                    String fecha = partes[1];
                    char estado = partes[2].charAt(0);

                    String textoEstado = (estado == 'A') ? "Asistio" : "Falto";

                    if (estado == 'F') {
                        totalFaltas++;
                    }

                    System.out.printf("  %-22s %-15s %-10s\n", nombre, fecha, textoEstado);
                    totalRegistros++;
                }
            }
            
            System.out.println("=========================================================");
            System.out.println(" Total de registros procesados: " + totalRegistros);
            System.out.println(" Total de FALTAS contabilizadas: " + totalFaltas);
            System.out.println("=========================================================");

        } catch (IOException e) {
            System.out.println("Error al intentar leer el archivo: " + e.getMessage());
        }
    }

    // ============================================================================
    // PROGRAMA PRINCIPAL
// ============================================================================
    public static void main(String[] args) {
        int opcion = 0;

        // Ciclo principal del menú de control de asistencia
        do {
            System.out.println("\n==================================");
            System.out.println("    SISTEMA CONTROL DE ASISTENCIA ");
            System.out.println("==================================");
            System.out.println("1. Registrar Asistencia");
            System.out.println("2. Mostrar Historial y Faltas");
            System.out.println("3. Salir del programa");
            System.out.println("----------------------------------");
            System.out.print("Seleccione una opcion (1-3): ");

            // Validación para evitar bucles infinitos por ingreso de letras
            if (teclado.hasNextInt()) {
                opcion = teclado.nextInt();
                
                switch (opcion) {
                    case 1:
                        registrarAsistencia();
                        break;
                    case 2:
                        mostrarHistorialYFaltas();
                        break;
                    case 3:
                        System.out.println("\nSaliendo del sistema de asistencia... ¡Hasta luego!");
                        break;
                    default:
                        System.out.println("[Error] La opcion elegida no existe. Intente con 1, 2 o 3.");
                }
            } else {
                System.out.println("[Error] Entrada no valida. Debe ingresar un numero entero.");
                teclado.next(); // Descarta la entrada incorrecta del buffer
                opcion = 0; // Reinicia la opción para que continúe el ciclo
            }
            
        } while (opcion != 3);

        teclado.close(); // Cerramos el recurso del teclado
    }
}
